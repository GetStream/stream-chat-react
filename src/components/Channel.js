import React, { PureComponent } from 'react';
import { withChatContext } from '../context';

import { LoadingIndicator } from './LoadingIndicator';

import PropTypes from 'prop-types';
import { MessageSimple } from './MessageSimple';
import { Attachment } from './Attachment';
import { ChannelInner } from './ChannelInner';

/**
 * Channel - Wrapper component for a channel. It needs to be place inside of the Chat component.
 * ChannelHeader, MessageList, Thread and MessageInput should be used as children of the Channel component.
 *
 * @example ./docs/Channel.md
 * @extends PureComponent
 */
class Channel extends PureComponent {
  constructor(props) {
    super(props);
    this.state = { error: false };
  }
  static propTypes = {
    /** Which channel to connect to, will initialize the channel if it's not initialized yet */
    channel: PropTypes.shape({
      watch: PropTypes.func,
    }).isRequired,
    /** Client is passed automatically via the Chat Context */
    client: PropTypes.object.isRequired,
    /** The loading indicator to use */
    LoadingIndicator: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),
    Message: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),
    Attachment: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),
  };

  static defaultProps = {
    LoadingIndicator,
    Message: MessageSimple,
    Attachment,
  };

  render() {
    if (!this.props.channel.cid) {
      return null; // <div>Select a channel</div>;
    }
    // We use a wrapper to make sure the key variable is set.
    // this ensures that if you switch channel the component is recreated
    return <ChannelInner {...this.props} key={this.props.channel.cid} />;
  }
}

Channel = withChatContext(Channel);

export { Channel };
