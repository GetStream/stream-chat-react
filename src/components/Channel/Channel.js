// @ts-check
/* eslint-disable sonarjs/no-duplicate-string */
import React, {
  useEffect,
  useCallback,
  useContext,
  useRef,
  useReducer,
  useLayoutEffect,
} from 'react';
import { v4 as uuidv4 } from 'uuid';
import PropTypes from 'prop-types';
import debounce from 'lodash.debounce';
import throttle from 'lodash.throttle';
import { logChatPromiseExecution, Channel as StreamChannel } from 'stream-chat';

import { ChatContext, ChannelContext, TranslationContext } from '../../context';
import { Attachment as AttachmentComponent } from '../Attachment';
import { MessageSimple as MessageSimpleComponent } from '../Message';
import {
  LoadingIndicator as LoadingIndicatorComponent,
  LoadingErrorIndicator as LoadingErrorIndicatorComponent,
} from '../Loading';
import useMentionsHandlers from './hooks/useMentionsHandlers';
import useEditMessageHandler from './hooks/useEditMessageHandler';
import useIsMounted from './hooks/useIsMounted';
import { channelReducer, initialState } from './channelState';

/** @type {React.FC<import('types').ChannelProps>}>} */
const Channel = ({ EmptyPlaceholder = null, ...props }) => {
  const { channel: contextChannel } = useContext(ChatContext);
  const channel = props.channel || contextChannel;
  if (!channel?.cid) {
    return EmptyPlaceholder;
  }
  return <ChannelInner {...props} channel={channel} key={channel.cid} />;
};

Channel.defaultProps = {
  multipleUploads: true,
};

Channel.propTypes = {
  /** Which channel to connect to, will initialize the channel if it's not initialized yet */
  channel: PropTypes.instanceOf(StreamChannel),
  /**
   * Empty channel UI component. This will be shown on the screen if there is no active channel.
   *
   * Defaults to null which skips rendering the Channel
   *
   * */
  EmptyPlaceholder: PropTypes.element,
  /**
   * Error indicator UI component. This will be shown on the screen if channel query fails.
   *
   * Defaults to and accepts same props as: [LoadingErrorIndicator](https://getstream.github.io/stream-chat-react/#loadingerrorindicator)
   *
   * */
  // @ts-ignore elementType
  LoadingErrorIndicator: PropTypes.elementType,
  /**
   * Loading indicator UI component. This will be shown on the screen until the messages are
   * being queried from channel≈ì. Once the messages are loaded, loading indicator is removed from the screen
   * and replaced with children of the Channel component.
   *
   * Defaults to and accepts same props as: [LoadingIndicator](https://github.com/GetStream/stream-chat-react/blob/master/src/components/LoadingIndicator.js)
   */
  // @ts-ignore elementType
  LoadingIndicator: PropTypes.elementType,
  /**
   * Message UI component to display a message in message list.
   *
   * Available built-in components (also accepts the same props as):
   *
   * 1. [MessageSimple](https://github.com/GetStream/stream-chat-react/blob/master/src/components/MessageSimple.js) (default)
   * 2. [MessageTeam](https://github.com/GetStream/stream-chat-react/blob/master/src/components/MessageTeam.js)
   * 3. [MessageLivestream](https://github.com/GetStream/stream-chat-react/blob/master/src/components/MessageLivestream.js)
   * 3. [MessageCommerce](https://github.com/GetStream/stream-chat-react/blob/master/src/components/MessageCommerce.js)
   *
   * */
  // @ts-ignore elementType
  Message: PropTypes.elementType,
  /**
   * Attachment UI component to display attachment in individual message.
   *
   * Defaults to and accepts same props as: [Attachment](https://github.com/GetStream/stream-chat-react/blob/master/src/components/Attachment.js)
   * */
  // @ts-ignore elementType
  Attachment: PropTypes.elementType,
  /**
   * Handle for click on @mention in message
   *
   * @param {Event} event DOM Click event
   * @param {User} user   Target [user object](https://getstream.io/chat/docs/#chat-doc-set-user) which is clicked
   */
  onMentionsClick: PropTypes.func,
  /**
   * Handle for hover on @mention in message
   *
   * @param {Event} event DOM hover event
   * @param {User} user   Target [user object](https://getstream.io/chat/docs/#chat-doc-set-user) which is hovered
   */
  onMentionsHover: PropTypes.func,
  /** Whether to allow multiple attachment uploads */
  multipleUploads: PropTypes.bool,
  /** List of accepted file types */
  acceptedFiles: PropTypes.array,
  /** Maximum number of attachments allowed per message */
  maxNumberOfFiles: PropTypes.number,
  /** Override send message request (Advanced usage only)
   *
   * @param {String} channelId full channel ID in format of `type:id`
   * @param {Object} message
   */
  doSendMessageRequest: PropTypes.func,
  /**
   * Override mark channel read request (Advanced usage only)
   *
   * @param {Channel} channel object
   * */
  doMarkReadRequest: PropTypes.func,
  /** Override update(edit) message request (Advanced usage only)
   *
   * @param {String} channelId full channel ID in format of `type:id`
   * @param {Object} updatedMessage
   */
  doUpdateMessageRequest: PropTypes.func,
};

/** @type {React.FC<import('types').ChannelProps & { channel: import('stream-chat').Channel }>} */
const ChannelInner = ({
  LoadingIndicator = LoadingIndicatorComponent,
  LoadingErrorIndicator = LoadingErrorIndicatorComponent,
  Attachment = AttachmentComponent,
  Message = MessageSimpleComponent,
  doMarkReadRequest,
  ...props
  // eslint-disable-next-line sonarjs/cognitive-complexity
}) => {
  const { channel } = props;
  const [state, dispatch] = useReducer(channelReducer, initialState);
  const originalTitle = useRef('');
  const lastRead = useRef(new Date());
  const chatContext = useContext(ChatContext);
  const online = useRef(true);
  const isMounted = useIsMounted();
  const { t } = useContext(TranslationContext);

  // eslint-disable-next-line react-hooks/exhaustive-deps
  const throttledCopyStateFromChannel = useCallback(
    throttle(
      () => {
        dispatch({ type: 'copyStateFromChannelOnEvent', channel });
      },
      500,
      { leading: true, trailing: true },
    ),
    [channel],
  );

  const markRead = useCallback(() => {
    if (channel.disconnected || !channel.getConfig()?.read_events) {
      return;
    }
    lastRead.current = new Date();
    if (doMarkReadRequest) {
      doMarkReadRequest(channel);
    } else {
      logChatPromiseExecution(channel.markRead(), 'mark read');
    }
    if (originalTitle.current) {
      document.title = originalTitle.current;
    }
  }, [channel, doMarkReadRequest]);

  // eslint-disable-next-line react-hooks/exhaustive-deps
  const markReadThrottled = useCallback(
    throttle(markRead, 500, { leading: true, trailing: true }),
    [markRead],
  );

  const handleEvent = useCallback(
    (e) => {
      dispatch({ type: 'updateThreadOnEvent', message: e.message, channel });
      if (e.type === 'connection.changed') {
        online.current = e.online;
      }

      if (e.type === 'message.new') {
        let mainChannelUpdated = true;
        if (e.message.parent_id && !e.message.show_in_channel) {
          mainChannelUpdated = false;
        }

        if (
          mainChannelUpdated &&
          e.message.user.id !== chatContext.client.userID
        ) {
          if (!document.hidden) {
            markReadThrottled();
          } else if (
            channel.getConfig()?.read_events &&
            !channel.muteStatus().muted
          ) {
            const unread = channel.countUnread(lastRead.current);
            document.title = `(${unread}) ${originalTitle.current}`;
          }
        }
      }

      throttledCopyStateFromChannel();
    },
    [
      channel,
      throttledCopyStateFromChannel,
      chatContext.client.userID,
      markReadThrottled,
    ],
  );

  // useLayoutEffect here to prevent spinner. Use Suspense when it is available in stable release
  useLayoutEffect(() => {
    let errored = false;
    let done = false;
    const onVisibilityChange = () => {
      if (!document.hidden) {
        markRead();
      }
    };

    (async () => {
      if (!channel.initialized) {
        try {
          await channel.watch();
        } catch (e) {
          dispatch({ type: 'setError', error: e });
          errored = true;
        }
      }
      done = true;
      originalTitle.current = document.title;
      if (!errored) {
        dispatch({ type: 'initStateFromChannel', channel });
        if (channel.countUnread() > 0) markRead();
        // The more complex sync logic is done in chat.js
        // listen to client.connection.recovered and all channel events
        document.addEventListener('visibilitychange', onVisibilityChange);
        chatContext.client.on('connection.changed', handleEvent);
        chatContext.client.on('connection.recovered', handleEvent);
        channel.on(handleEvent);
      }
    })();
    return () => {
      if (errored || !done) return;
      document.removeEventListener('visibilitychange', onVisibilityChange);
      channel.off(handleEvent);
      chatContext.client.off('connection.changed', handleEvent);
      chatContext.client.off('connection.recovered', handleEvent);
    };
  }, [channel, chatContext.client, handleEvent, markRead, props.channel]);

  useEffect(() => {
    if (state.thread) {
      for (let i = state.messages.length - 1; i >= 0; i -= 1) {
        if (state.messages[i].id === state.thread.id) {
          dispatch({ type: 'setThread', message: state.messages[i] });
          break;
        }
      }
    }
  }, [state.messages, state.thread]);

  // Message
  // eslint-disable-next-line react-hooks/exhaustive-deps
  const loadMoreFinished = useCallback(
    debounce(
      /**
       * @param {boolean} hasMore
       * @param {import('stream-chat').ChannelState['messages']} messages
       */
      (hasMore, messages) => {
        if (!isMounted.current) return;
        dispatch({ type: 'loadMoreFinished', hasMore, messages });
      },
      2000,
      {
        leading: true,
        trailing: true,
      },
    ),
    [],
  );

  const loadMore = useCallback(
    async (limit = 100) => {
      if (!online.current || !window.navigator.onLine) return 0;
      // prevent duplicate loading events...
      const oldestMessage = state.messages[0];
      if (state.loadingMore || oldestMessage?.status !== 'received') return 0;
      dispatch({ type: 'setLoadingMore', loadingMore: true });

      const oldestID = oldestMessage?.id;

      const perPage = limit;
      let queryResponse;
      try {
        queryResponse = await channel.query({
          messages: { limit: perPage, id_lt: oldestID },
        });
      } catch (e) {
        console.warn('message pagination request failed with error', e);
        dispatch({ type: 'setLoadingMore', loadingMore: false });
        return 0;
      }
      const hasMoreMessages = queryResponse.messages.length === perPage;

      loadMoreFinished(hasMoreMessages, channel.state.messages);

      return queryResponse.messages.length;
    },
    [channel, loadMoreFinished, state.loadingMore, state.messages, online],
  );

  const updateMessage = useCallback(
    (updatedMessage) => {
      // adds the message to the local channel state..
      // this adds to both the main channel state as well as any reply threads
      channel.state.addMessageSorted(updatedMessage, true);

      dispatch({
        type: 'copyMessagesFromChannel',
        parentId: state.thread && updatedMessage.parent_id,
        channel,
      });
    },
    [channel, state.thread],
  );

  const { doSendMessageRequest } = props;
  const doSendMessage = useCallback(
    async (message) => {
      const { text, attachments, id, parent_id, mentioned_users } = message;
      const messageData = {
        text,
        attachments,
        mentioned_users,
        id,
        parent_id,
      };

      try {
        let messageResponse;
        if (doSendMessageRequest) {
          messageResponse = await doSendMessageRequest(
            channel.cid,
            messageData,
          );
        } else {
          messageResponse = await channel.sendMessage(messageData);
        }

        // replace it after send is completed
        if (messageResponse && messageResponse.message) {
          updateMessage({
            ...messageResponse.message,
            status: 'received',
          });
        }
      } catch (e) {
        // set the message to failed..
        updateMessage({
          ...message,
          status: 'failed',
        });
      }
    },
    [channel, doSendMessageRequest, updateMessage],
  );

  const createMessagePreview = useCallback(
    (text, attachments, parent, mentioned_users) => {
      // create a preview of the message
      const clientSideID = `${chatContext.client.userID}-${uuidv4()}`;
      return {
        text,
        html: text,
        __html: text,
        id: clientSideID,
        type: 'regular',
        status: 'sending',
        user: chatContext.client.user,
        created_at: new Date(),
        attachments,
        mentioned_users,
        reactions: [],
        ...(parent?.id ? { parent_id: parent.id } : null),
      };
    },
    [chatContext.client.user, chatContext.client.userID],
  );

  const sendMessage = useCallback(
    async ({ text, attachments = [], mentioned_users = [], parent }) => {
      // remove error messages upon submit
      channel.state.filterErrorMessages();

      // create a local preview message to show in the UI
      const messagePreview = createMessagePreview(
        text,
        attachments,
        parent,
        mentioned_users,
      );

      // first we add the message to the UI
      updateMessage(messagePreview);

      await doSendMessage(messagePreview);
    },
    [channel.state, createMessagePreview, doSendMessage, updateMessage],
  );

  const retrySendMessage = useCallback(
    async (message) => {
      // set the message status to sending
      updateMessage({
        ...message,
        status: 'sending',
      });
      // actually try to send the message...
      await doSendMessage(message);
    },
    [doSendMessage, updateMessage],
  );

  const removeMessage = useCallback(
    (message) => {
      channel.state.removeMessage(message);
      dispatch({
        type: 'copyMessagesFromChannel',
        parentId: state.thread && message.parent_id,
        channel,
      });
    },
    [channel, state.thread],
  );

  // Thread

  const openThread = useCallback(
    (message, e) => {
      if (e && e.preventDefault) {
        e.preventDefault();
      }

      dispatch({ type: 'openThread', message, channel });
    },
    [channel],
  );

  // eslint-disable-next-line react-hooks/exhaustive-deps
  const loadMoreThreadFinished = useCallback(
    debounce(
      /**
       * @param {boolean} threadHasMore
       * @param {import('seamless-immutable').ImmutableArray<ReturnType<import('stream-chat').ChannelState['messageToImmutable']>>} threadMessages
       */
      (threadHasMore, threadMessages) => {
        dispatch({
          type: 'loadMoreThreadFinished',
          threadHasMore,
          threadMessages,
        });
      },
      2000,
      { leading: true, trailing: true },
    ),
    [],
  );

  const loadMoreThread = useCallback(async () => {
    // prevent duplicate loading events...
    if (state.threadLoadingMore || !state.thread) return;
    dispatch({ type: 'startLoadingThread' });
    const parentID = state.thread.id;

    if (!parentID) {
      dispatch({ type: 'closeThread' });
      return;
    }
    const oldMessages = channel.state.threads[parentID] || [];
    const oldestMessageID = oldMessages[0]?.id;
    const limit = 50;
    const queryResponse = await channel.getReplies(parentID, {
      limit,
      id_lt: oldestMessageID,
    });

    const threadHasMoreMessages = queryResponse.messages.length === limit;

    const newThreadMessages = channel.state.threads[parentID] || [];

    // next set loadingMore to false so we can start asking for more data...
    loadMoreThreadFinished(threadHasMoreMessages, newThreadMessages);
  }, [channel, loadMoreThreadFinished, state.thread, state.threadLoadingMore]);

  const closeThread = useCallback((e) => {
    if (e && e.preventDefault) {
      e.preventDefault();
    }
    dispatch({ type: 'closeThread' });
  }, []);

  const onMentionsHoverOrClick = useMentionsHandlers(
    props.onMentionsHover,
    props.onMentionsClick,
  );

  const editMessage = useEditMessageHandler(props.doUpdateMessageRequest);

  const channelContextValue = {
    // state
    ...state,
    watcher_count: state.watcherCount,
    // props
    channel,
    Message,
    Attachment,
    multipleUploads: props.multipleUploads,
    acceptedFiles: props.acceptedFiles,
    maxNumberOfFiles: props.maxNumberOfFiles,
    mutes: chatContext.mutes,
    // handlers
    loadMore,
    editMessage,
    updateMessage,
    sendMessage,
    retrySendMessage,
    removeMessage,
    openThread,
    loadMoreThread,
    closeThread,
    onMentionsClick: onMentionsHoverOrClick,
    onMentionsHover: onMentionsHoverOrClick,
    // from chatContext, for legacy reasons
    client: chatContext.client,
  };

  let core;
  if (state.error) {
    core = <LoadingErrorIndicator error={state.error} />;
  } else if (state.loading) {
    core = <LoadingIndicator size={25} />;
  } else if (!props.channel?.watch) {
    core = <div>{t('Channel Missing')}</div>;
  } else {
    core = (
      <ChannelContext.Provider value={channelContextValue}>
        <div className="str-chat__container">{props.children}</div>
      </ChannelContext.Provider>
    );
  }

  return (
    <div className={`str-chat str-chat-channel ${chatContext.theme}`}>
      {core}
    </div>
  );
};

export default React.memo(Channel);
