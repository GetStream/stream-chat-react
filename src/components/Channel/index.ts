export * from './Channel';
export { useEditMessageHandler as useChannelEditMessageHandler } from './hooks/useEditMessageHandler';
export { useMentionsHandlers as useChannelMentionsHandler } from './hooks/useMentionsHandlers';
export { useChannelContainerClasses } from './hooks/useChannelContainerClasses';
export { makeAddNotifications } from './utils';
