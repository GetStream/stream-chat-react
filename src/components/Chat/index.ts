export * from './Chat';
export * from './hooks/useChat';
export * from './hooks/useCustomStyles';
export * from './hooks/useCreateChatClient';
