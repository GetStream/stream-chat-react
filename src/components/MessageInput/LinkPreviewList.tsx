import clsx from 'clsx';
import React, { useState } from 'react';
import { useChannelStateContext, useMessageInputContext } from '../../context';
import type { LinkPreview } from './types';
import { LinkPreviewState } from './types';
import { CloseIcon, LinkIcon } from './icons';
import { PopperTooltip } from '../Tooltip';
import { useEnterLeaveHandlers } from '../Tooltip/hooks';

export type LinkPreviewListProps = {
  linkPreviews: LinkPreview[];
};

export const LinkPreviewList = ({ linkPreviews }: LinkPreviewListProps) => {
  const { quotedMessage } = useChannelStateContext();
  const showLinkPreviews = linkPreviews.length > 0 && !quotedMessage;

  if (!showLinkPreviews) return null;

  return (
    <div className='str-chat__link-preview-list'>
      {Array.from(linkPreviews.values()).map((linkPreview) =>
        linkPreview.state === LinkPreviewState.LOADED ? (
          <LinkPreviewCard key={linkPreview.og_scrape_url} linkPreview={linkPreview} />
        ) : null,
      )}
    </div>
  );
};

type LinkPreviewProps = {
  linkPreview: LinkPreview;
};

const LinkPreviewCard = ({ linkPreview }: LinkPreviewProps) => {
  const { dismissLinkPreview } = useMessageInputContext();
  const { handleEnter, handleLeave, tooltipVisible } = useEnterLeaveHandlers<HTMLDivElement>();
  const [referenceElement, setReferenceElement] = useState<HTMLDivElement | null>(null);
  return (
    <div
      className={clsx('str-chat__link-preview-card', {
        'str-chat__link-preview-card--loading': linkPreview.state === LinkPreviewState.LOADING,
      })}
      data-testid='link-preview-card'
    >
      <PopperTooltip offset={[0, 5]} referenceElement={referenceElement} visible={tooltipVisible}>
        {linkPreview.og_scrape_url}
      </PopperTooltip>
      <div
        className='str-chat__link-preview-card__icon-container'
        onMouseEnter={handleEnter}
        onMouseLeave={handleLeave}
        ref={setReferenceElement}
      >
        <LinkIcon />
      </div>
      <div className='str-chat__link-preview-card__content'>
        <div className='str-chat__link-preview-card__content-title'>{linkPreview.title}</div>
        <div className='str-chat__link-preview-card__content-description'>{linkPreview.text}</div>
      </div>
      <button
        className='str-chat__link-preview-card__dismiss-button'
        data-testid='link-preview-card-dismiss-btn'
        onClick={() => dismissLinkPreview(linkPreview)}
      >
        <CloseIcon />
      </button>
    </div>
  );
};
