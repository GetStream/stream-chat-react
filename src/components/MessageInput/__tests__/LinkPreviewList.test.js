import {
  generateChannel,
  generateMember,
  generateMessage,
  generateScrapedImageAttachment,
  generateUser,
  initClientWithChannels,
} from '../../../mock-builders';
import { act, cleanup, fireEvent, render, screen } from '@testing-library/react';
import React from 'react';
import { Chat } from '../../Chat';
import { Channel } from '../../Channel';

import '@testing-library/jest-dom';
import { LinkPreviewStatus } from 'stream-chat';
import { LinkPreviewCard, LinkPreviewList } from '../LinkPreviewList';

const LINK_PREVIEW_TEST_ID = 'link-preview-card';
const userId = 'userId';
const username = 'username';
const mentionId = 'mention-id';
const mentionName = 'mention-name';
const user1 = generateUser({ id: userId, name: username });
const mentionUser = generateUser({
  id: mentionId,
  name: mentionName,
});
const mainListMessage = generateMessage({ user: user1 });
const threadMessage = generateMessage({
  parent_id: mainListMessage.id,
  type: 'reply',
  user: user1,
});
const mockedChannelData = generateChannel({
  members: [generateMember({ user: user1 }), generateMember({ user: mentionUser })],
  messages: [mainListMessage],
  thread: [threadMessage],
});

const renderComponent = async ({ channelProps = {}, client } = {}) => {
  let renderResult;
  await act(() => {
    renderResult = render(
      <Chat client={client}>
        <Channel {...channelProps}>
          <LinkPreviewList />
        </Channel>
      </Chat>,
    );
  });
  const submit = async () => {
    const submitButton =
      renderResult.findByText('Send') || renderResult.findByTitle('Send');
    fireEvent.click(await submitButton);
  };

  return { submit, ...renderResult };
};

const setup = async () => {
  const {
    channels: [channel],
    client,
  } = await initClientWithChannels({
    channelsData: [mockedChannelData],
  });
  return { channel, client };
};

describe('LinkPreviewList', () => {
  afterEach(() => {
    cleanup();
    jest.clearAllMocks();
  });

  const previews = Array.from({ length: 4 })
    .map(() => generateScrapedImageAttachment({ status: LinkPreviewStatus.LOADED }))
    .reduce((acc, p) => {
      acc.set(p.og_scrape_url, p);
      return acc;
    }, new Map());

  it('renders previews for each link', async () => {
    const { channel, client } = await setup();
    await renderComponent({
      channelProps: { channel },
      client,
    });
    await act(() => {
      channel.messageComposer.linkPreviewsManager.state.next({ previews });
    });
    const linkPreviewCards = screen.getAllByTestId(LINK_PREVIEW_TEST_ID);
    expect(linkPreviewCards).toHaveLength(previews.size);
    previews.values().forEach((p, i) => {
      expect(linkPreviewCards[i]).toHaveTextContent(p.title);
    });
  });
  it('does not render if no previews are available', async () => {
    const previews = new Map();
    const { channel, client } = await setup();
    await renderComponent({
      channelProps: { channel },
      client,
    });
    await act(() => {
      channel.messageComposer.linkPreviewsManager.state.next({ previews });
    });
    const linkPreviewCards = screen.queryAllByTestId(LINK_PREVIEW_TEST_ID);
    expect(linkPreviewCards).toHaveLength(0);
  });

  it('does not render if quoting a message', async () => {
    const { channel, client } = await setup();
    await renderComponent({
      channelProps: { channel },
      client,
    });
    await act(() => {
      channel.messageComposer.linkPreviewsManager.state.next({ previews });
      channel.messageComposer.state.next({ quotedMessage: generateMessage() });
    });
    const linkPreviewCards = screen.queryAllByTestId(LINK_PREVIEW_TEST_ID);
    expect(linkPreviewCards).toHaveLength(0);
  });
});

const renderLinkPreviewCard = async ({ channel, client, linkPreview }) => {
  let result;
  await act(() => {
    result = render(
      <Chat client={client}>
        <Channel channel={channel}>
          <LinkPreviewCard linkPreview={linkPreview} />
        </Channel>
      </Chat>,
    );
  });
  return result;
};
describe('LinPreviewCard', () => {
  it('renders for loaded preview', async () => {
    const { channel, client } = await setup();
    const { container } = await renderLinkPreviewCard({
      channel,
      client,
      linkPreview: generateScrapedImageAttachment({
        og_scrape_url: 'og_scrape_url',
        status: LinkPreviewStatus.LOADED,
        text: 'text',
        title: 'title',
      }),
    });
    expect(container).toMatchSnapshot();
  });
  it('renders for loading preview', async () => {
    const { channel, client } = await setup();
    const { container } = await renderLinkPreviewCard({
      channel,
      client,
      linkPreview: generateScrapedImageAttachment({
        og_scrape_url: 'og_scrape_url',
        status: LinkPreviewStatus.LOADING,
        text: 'text',
        title: 'title',
      }),
    });
    expect(container).toMatchSnapshot();
  });
  it('does not render dismissed preview', async () => {
    const { channel, client } = await setup();
    const { container } = await renderLinkPreviewCard({
      channel,
      client,
      linkPreview: generateScrapedImageAttachment({
        og_scrape_url: 'og_scrape_url',
        status: LinkPreviewStatus.DISMISSED,
        text: 'text',
        title: 'title',
      }),
    });
    expect(container).toMatchSnapshot();
  });
  it('does not render failed preview', async () => {
    const { channel, client } = await setup();
    const { container } = await renderLinkPreviewCard({
      channel,
      client,
      linkPreview: generateScrapedImageAttachment({
        og_scrape_url: 'og_scrape_url',
        status: LinkPreviewStatus.FAILED,
        text: 'text',
        title: 'title',
      }),
    });
    expect(container).toMatchSnapshot();
  });
  it('does not render pending preview', async () => {
    const { channel, client } = await setup();
    const { container } = await renderLinkPreviewCard({
      channel,
      client,
      linkPreview: generateScrapedImageAttachment({
        og_scrape_url: 'og_scrape_url',
        status: LinkPreviewStatus.PENDING,
        text: 'text',
        title: 'title',
      }),
    });
    expect(container).toMatchSnapshot();
  });
  it('allows to dismiss a preview', async () => {
    const { channel, client } = await setup();
    const dismissPreviewMock = jest.spyOn(
      channel.messageComposer.linkPreviewsManager,
      'dismissPreview',
    );
    await await renderLinkPreviewCard({
      channel,
      client,
      linkPreview: generateScrapedImageAttachment({ status: LinkPreviewStatus.LOADED }),
    });

    fireEvent.click(screen.getByTestId('link-preview-card-dismiss-btn'));
    expect(dismissPreviewMock).toHaveBeenCalled();
  });
});
