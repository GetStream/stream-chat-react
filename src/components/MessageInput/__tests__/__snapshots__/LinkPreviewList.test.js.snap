// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LinPreviewCard does not render dismissed preview 1`] = `
<div>
  <div
    class="str-chat messaging light str-chat__channel"
    id="str-chat__channel"
  >
    <div
      class="str-chat__container"
    />
  </div>
</div>
`;

exports[`LinPreviewCard does not render failed preview 1`] = `
<div>
  <div
    class="str-chat messaging light str-chat__channel"
    id="str-chat__channel"
  >
    <div
      class="str-chat__container"
    />
  </div>
</div>
`;

exports[`LinPreviewCard does not render pending preview 1`] = `
<div>
  <div
    class="str-chat messaging light str-chat__channel"
    id="str-chat__channel"
  >
    <div
      class="str-chat__container"
    />
  </div>
</div>
`;

exports[`LinPreviewCard renders for loaded preview 1`] = `
<div>
  <div
    class="str-chat messaging light str-chat__channel"
    id="str-chat__channel"
  >
    <div
      class="str-chat__container"
    >
      <div
        class="str-chat__link-preview-card"
        data-testid="link-preview-card"
      >
        <div
          class="str-chat__link-preview-card__icon-container"
        >
          <svg
            fill="none"
            height="11"
            viewBox="0 0 20 11"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M1.9 5.5C1.9 3.79 3.29 2.4 5 2.4H8.05C8.57467 2.4 9 1.97467 9 1.45C9 0.925329 8.57467 0.5 8.05 0.5H5C2.24 0.5 0 2.74 0 5.5C0 8.26 2.24 10.5 5 10.5H8.05C8.57467 10.5 9 10.0747 9 9.55C9 9.02533 8.57467 8.6 8.05 8.6H5C3.29 8.6 1.9 7.21 1.9 5.5ZM6 5.5C6 6.05228 6.44772 6.5 7 6.5H13C13.5523 6.5 14 6.05228 14 5.5C14 4.94772 13.5523 4.5 13 4.5H7C6.44772 4.5 6 4.94772 6 5.5ZM15 0.5H11.95C11.4253 0.5 11 0.925329 11 1.45C11 1.97467 11.4253 2.4 11.95 2.4H15C16.71 2.4 18.1 3.79 18.1 5.5C18.1 7.21 16.71 8.6 15 8.6H11.95C11.4253 8.6 11 9.02533 11 9.55C11 10.0747 11.4253 10.5 11.95 10.5H15C17.76 10.5 20 8.26 20 5.5C20 2.74 17.76 0.5 15 0.5Z"
              fill="#005DFF"
              fill-rule="evenodd"
            />
          </svg>
        </div>
        <div
          class="str-chat__link-preview-card__content"
        >
          <div
            class="str-chat__link-preview-card__content-title"
          >
            title
          </div>
          <div
            class="str-chat__link-preview-card__content-description"
          >
            text
          </div>
        </div>
        <button
          class="str-chat__link-preview-card__dismiss-button"
          data-testid="link-preview-card-dismiss-btn"
          type="button"
        >
          <svg
            data-testid="close-no-outline"
            fill="none"
            height="24"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M19 6.41L17.59 5L12 10.59L6.41 5L5 6.41L10.59 12L5 17.59L6.41 19L12 13.41L17.59 19L19 17.59L13.41 12L19 6.41Z"
              fill="black"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`LinPreviewCard renders for loading preview 1`] = `
<div>
  <div
    class="str-chat messaging light str-chat__channel"
    id="str-chat__channel"
  >
    <div
      class="str-chat__container"
    >
      <div
        class="str-chat__link-preview-card str-chat__link-preview-card--loading"
        data-testid="link-preview-card"
      >
        <div
          class="str-chat__link-preview-card__icon-container"
        >
          <svg
            fill="none"
            height="11"
            viewBox="0 0 20 11"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M1.9 5.5C1.9 3.79 3.29 2.4 5 2.4H8.05C8.57467 2.4 9 1.97467 9 1.45C9 0.925329 8.57467 0.5 8.05 0.5H5C2.24 0.5 0 2.74 0 5.5C0 8.26 2.24 10.5 5 10.5H8.05C8.57467 10.5 9 10.0747 9 9.55C9 9.02533 8.57467 8.6 8.05 8.6H5C3.29 8.6 1.9 7.21 1.9 5.5ZM6 5.5C6 6.05228 6.44772 6.5 7 6.5H13C13.5523 6.5 14 6.05228 14 5.5C14 4.94772 13.5523 4.5 13 4.5H7C6.44772 4.5 6 4.94772 6 5.5ZM15 0.5H11.95C11.4253 0.5 11 0.925329 11 1.45C11 1.97467 11.4253 2.4 11.95 2.4H15C16.71 2.4 18.1 3.79 18.1 5.5C18.1 7.21 16.71 8.6 15 8.6H11.95C11.4253 8.6 11 9.02533 11 9.55C11 10.0747 11.4253 10.5 11.95 10.5H15C17.76 10.5 20 8.26 20 5.5C20 2.74 17.76 0.5 15 0.5Z"
              fill="#005DFF"
              fill-rule="evenodd"
            />
          </svg>
        </div>
        <div
          class="str-chat__link-preview-card__content"
        >
          <div
            class="str-chat__link-preview-card__content-title"
          >
            title
          </div>
          <div
            class="str-chat__link-preview-card__content-description"
          >
            text
          </div>
        </div>
        <button
          class="str-chat__link-preview-card__dismiss-button"
          data-testid="link-preview-card-dismiss-btn"
          type="button"
        >
          <svg
            data-testid="close-no-outline"
            fill="none"
            height="24"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M19 6.41L17.59 5L12 10.59L6.41 5L5 6.41L10.59 12L5 17.59L6.41 19L12 13.41L17.59 19L19 17.59L13.41 12L19 6.41Z"
              fill="black"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;
